main.go

```diff
@@ -14,47 +16,220 @@ import (
+			if i < len(v.Latencies)+offset && v.Latencies[i-offset].Destination == u {
+				nV.Latencies[i] = v.Latencies[i-offset]
+				continue
+			}
+			offset++
+			nV.Latencies[i].Destination = "dummy"
+			nV.Latencies[i].Ok = true
+		}
+		nm[k] = nV
+	}
+	return nm
+}
+
+func ipOrHost(ip, host string) string {
+	if ip != "na" {
+		return ip
+	}
+	return host
+}
+
+func (m matrix) String(f format) string {
+	tableString := &strings.Builder{}
+	table := tablewriter.NewWriter(tableString)
+	var data [][]string
+	switch f {
+	case fancy:
+		line := []string{"Source\\Dest"}
+		for _, v := range m {
+			line = append(line, ipOrHost(v.IP, v.Host))
+		}
+		table.SetHeader(line)
+		line = []string{}
+		for _, v := range m {
+			line = []string{ipOrHost(v.IP, v.Host)}
+			for _, l := range v.Latencies {
+				line = append(line, fmt.Sprintf("%s code:%t", l.Duration.String(), l.Ok))
+			}
+			data = append(data, line)
 		}
-		s += "\n"
+		table.SetAutoFormatHeaders(true)
+		table.SetHeaderAlignment(tablewriter.ALIGN_LEFT)
+	case simple:
+		for _, v := range m {
+			line := []string{}
+			for _, l := range v.Latencies {
+				line = append(line, l.Duration.String())
+			}
+			data = append(data, line)
+		}
+		table.SetCenterSeparator("")
+		table.SetColumnSeparator("")
+		table.SetRowSeparator("")
+		table.SetHeaderLine(false)
+		table.SetBorder(false)
+	default:
+		for _, v := range m {
+			line := []string{}
+			for _, l := range v.Latencies {
+				line = append(line, fmt.Sprintf("%s code:%t", l.Duration.String(), l.Ok))
+			}
+			data = append(data, line)
+		}
+		table.SetCenterSeparator("")
+		table.SetColumnSeparator("")
+		table.SetRowSeparator("")
+		table.SetHeaderLine(false)
+		table.SetBorder(false)
 	}
-	return s
+	table.SetAlignment(tablewriter.ALIGN_LEFT)
+	table.SetTablePadding(" ")
+	table.AppendBulk(data)
+	table.Render()
+	return tableString.String()
 }
 
-func timeHTTPRequest(ctx context.Context, url string) *Latency {
-	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+func timeHTTPRequest(ctx context.Context, u *url.URL) *Latency {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
 	if err != nil {
-		return &Latency{Destination: url}
+		log.Printf("failed to create Request: %v\n", err)
+		errorCounter.Inc()
+		return &Latency{Host: u.Hostname(), Destination: u.String()}
 	}
 	start := time.Now()
-	if _, err := http.DefaultClient.Do(req); err != nil {
-		fmt.Printf("failed to make ping request: %v\n", err)
+	var end time.Time
+	if _, err = httpPingClient.Do(req); err != nil {
+		log.Printf("failed to make ping request: %v\n", err)
+		// set the time to almost infinity
+		end = time.Unix(1<<63-1, 0)
+	} else {
+		end = time.Now()
+	}
+	// Try to get IP address of target
+	// Shadow the err, because not being able to get an IP address should not
+	// overwrite the previous error and getting no error does not indicate, that
+	// the fake ping request was successful
+	ip := "na"
```

Comment:
If we are using this string as a sentinel value, then it should probably be a `const`